### PR TITLE
adding newsletter mailjet iframe, fix #11, fix #362

### LINKED ISSUE #11: #US19 - S'inscrire à la newsletter de cafébabel
ETQ visiteur (contributeurs + lecteurs) je veux pouvoir souscrire à la mailing list pour l'envoi de la newsletter afin d'avoir plus d'info sur cafébabel

- [x] - doit afficher un champs pour inscrire son adresse mail dans le footer
- [x] - doit permettre de valider la selection de l'adresse
- [x] - doit notifier que la souscription a bien été faite

### LINKED ISSUE #362: Possibility to subscribe to newsletter on the homepage
C'est un must

### PR DESCRIPTION
**Pull request by johangiraud**

### CODE REVIEW

File: cafebabel/static/js/script.js [hunk 1]
```diff
@@ -26,6 +26,50 @@ Array.from(document.querySelectorAll('a')).forEach(a => {
     a.setAttribute('target', '_blank')
 })
 
+/* smooth scroll to anchor https://stackoverflow.com/a/17733311/6481285 */
+function animate(elem, style, unit, from, to, time, prop) {
```
Comment (davidbgk) [inline cafebabel/static/js/script.js]: Remove parameters you don't change/use like `unit` and `prop` and simplify the code accordingly.

File: cafebabel/static/js/script.js [hunk 2]
```diff
@@ -26,6 +26,50 @@ Array.from(document.querySelectorAll('a')).forEach(a => {
...
+      var step = Math.min(1, (new Date().getTime() - start) / time)
+      if (prop) {
+        elem[style] = from + step * (to - from) + unit
+      } else {
+        elem.style[style] = from + step * (to - from) + unit
+      }
+      if (step === 1) {
+        clearInterval(timer)
+      }
+    }, 25)
```
Comment (davidbgk) [inline cafebabel/static/js/script.js]: Document magic numbers.

File: cafebabel/static/js/script.js [hunk 3]
```diff
@@ -26,6 +26,50 @@ Array.from(document.querySelectorAll('a')).forEach(a => {
...
+      ? window.pageYOffset
+      : (document.documentElement || document.body.parentNode || document.body)
+          .scrollTop
+  animate(
+    document.scrollingElement || document.documentElement,
+    'scrollTop',
+    '',
+    scrollPosition,
+    target.offsetTop,
+    300,
```
Comment (davidbgk) [inline cafebabel/static/js/script.js]: Same here.

**Review by davidbgk — CHANGES_REQUESTED** [review]

File: cafebabel/static/js/script.js (hunk 2 above)
Comment (johangiraud) [reply cafebabel/static/js/script.js]: you mean we don't know what `25` refers to? I don't :-(

File: cafebabel/static/js/script.js (hunk 3 above)
Comment (johangiraud) [reply cafebabel/static/js/script.js]: 300 is the transition duration, this one looks pretty clear to me

File: cafebabel/static/js/script.js (hunk 1 above)
Comment (johangiraud) [reply cafebabel/static/js/script.js]: ok ! https://github.com/cafebabel/cafebabel.com/pull/386/commits/18c6c62d86ce6ed12a0595491d309233d9f20c43

**Comment by vinyll** [comment]
@cafebabel: _WIP_ or to be reviewed?

**Comment by johangiraud** [comment]
Sorry: to be reviewed

File: cafebabel/static/js/script.js [hunk 4]
```diff
@@ -26,6 +26,40 @@ Array.from(document.querySelectorAll('a')).forEach(a => {
     a.setAttribute('target', '_blank')
 })
 
+/* smooth scroll to anchor https://stackoverflow.com/a/17733311/6481285 */
+function animate(elem, style, from, to, time) {
+  if (!elem) {
```
Comment (vinyll) [inline cafebabel/static/js/script.js]: one line is appropriate here `if (!elem) return` ;)

File: cafebabel/static/js/script.js [hunk 5]
```diff
@@ -26,6 +26,40 @@ Array.from(document.querySelectorAll('a')).forEach(a => {
...
+    return
+  }
+  var start = new Date().getTime(),
```
Comment (vinyll) [inline cafebabel/static/js/script.js]: use `const`

File: cafebabel/static/js/script.js [hunk 6]
```diff
@@ -26,6 +26,40 @@ Array.from(document.querySelectorAll('a')).forEach(a => {
...
+    timer = setInterval(() => {
```
Comment (vinyll) [inline cafebabel/static/js/script.js]: same here

File: cafebabel/static/js/script.js [hunk 7]
```diff
@@ -26,6 +26,40 @@ Array.from(document.querySelectorAll('a')).forEach(a => {
...
+      var step = Math.min(1, (new Date().getTime() - start) / time)
```
Comment (vinyll) [inline cafebabel/static/js/script.js]: here too

File: cafebabel/static/js/script.js [hunk 8]
```diff
@@ -26,6 +26,40 @@ Array.from(document.querySelectorAll('a')).forEach(a => {
...
+      elem[style] = from + step * (to - from)
+      if (step === 1) {
```
Comment (vinyll) [inline cafebabel/static/js/script.js]: you could eventually use a one liner here too. Up to you.

File: cafebabel/static/js/script.js (hunk 5 above)
Comment (davidbgk) [inline cafebabel/static/js/script.js]: `const start = new Date().getTime()`

File: cafebabel/static/js/script.js (hunk 6 above)
Comment (davidbgk) [inline cafebabel/static/js/script.js]: `const timer = setInterval(() => {` (separate definitions)

File: cafebabel/static/js/script.js [hunk 9]
```diff
@@ -26,6 +26,40 @@ Array.from(document.querySelectorAll('a')).forEach(a => {
...
+        clearInterval(timer)
+      }
+    }, 25)
```
Comment (davidbgk) [inline cafebabel/static/js/script.js]: `/* Delay in ms, set to 25 because… */`

File: cafebabel/static/js/script.js (hunk 7 above)
Comment (davidbgk) [inline cafebabel/static/js/script.js]: `const`

File: cafebabel/static/js/script.js [hunk 10]
```diff
@@ -26,6 +26,40 @@ Array.from(document.querySelectorAll('a')).forEach(a => {
...
+    window.pageYOffset !== undefined
+      ? window.pageYOffset
+      : (document.documentElement || document.body.parentNode || document.body)
+          .scrollTop
+  animate(
+    document.scrollingElement || document.documentElement,
+    'scrollTop',
+    scrollPosition,
+    target.offsetTop,
+    300
```
Comment (davidbgk) [inline cafebabel/static/js/script.js]: `/* Total(?) duration of the animation in ms, set to 300 because… */`

**Review by davidbgk — CHANGES_REQUESTED** [review]

File: cafebabel/static/js/script.js (hunk 9 above)
Comment (johangiraud) [reply cafebabel/static/js/script.js]: I removed it. Don't know why it was here

File: cafebabel/static/js/script.js (hunk 5 above)
Comment (johangiraud) [reply cafebabel/static/js/script.js]: ok

File: cafebabel/static/js/script.js (hunk 6 above)
Comment (johangiraud) [reply cafebabel/static/js/script.js]: ok

File: cafebabel/static/js/script.js (hunk 7 above)
Comment (johangiraud) [reply cafebabel/static/js/script.js]: ok

File: cafebabel/static/js/script.js (hunk 10 above)
Comment (johangiraud) [reply cafebabel/static/js/script.js]: haha I don't know how to justify 300, well 200 was too fast and 400 to slow I guess

File: cafebabel/templates/_social-networks-macro.html [hunk 11]
```diff
@@ -3,6 +3,7 @@
     <li><a href={{ social_network_url_for('facebook') }} class=icon-facebook></a></li>
     <li><a href={{ social_network_url_for('twitter') }} class=icon-twitter></a></li>
     <li><a href={{ social_network_url_for('instagram') }} class=icon-instagram></a></li>
+    <li><a href=#static-pages-links class=icon-newsletter></a></li>
```
Comment (vinyll) [inline cafebabel/templates/_social-networks-macro.html]: Funny I hadn't notify this!
A link with no content is totally inaccessible (I mean all these social items).
We should have text in HTML so it's readable for all reading systems, and remove it from display on screens.
Technically that mean the HTML is like 
```html
<li><a href={{ social_network_url_for('twitter') }} class=icon-twitter>Cafebabel on Twitter</a></li>
```
and a CSS trick would be:
```css
nav#socials > a {
  text-indent: -9999%;
}
```

**Review by vinyll — APPROVED** [review]
I'll approve not to be a blocker, but all the empty `<a>` in the nav are an issue.

File: cafebabel/static/js/script.js (hunk 7 above)
Comment (johangiraud) [reply cafebabel/static/js/script.js]: https://github.com/cafebabel/cafebabel.com/pull/386/commits/73063f4dfab0974f300604a87e748bd20cdea61c

File: cafebabel/static/js/script.js (hunk 8 above)
Comment (johangiraud) [reply cafebabel/static/js/script.js]: ok

File: cafebabel/static/js/script.js (hunk 6 above)
Comment (johangiraud) [reply cafebabel/static/js/script.js]: same https://github.com/cafebabel/cafebabel.com/pull/386/commits/73063f4dfab0974f300604a87e748bd20cdea61c

File: cafebabel/static/js/script.js (hunk 5 above)
Comment (johangiraud) [reply cafebabel/static/js/script.js]: already changed here https://github.com/cafebabel/cafebabel.com/pull/386/commits/73063f4dfab0974f300604a87e748bd20cdea61c

File: cafebabel/templates/_social-networks-macro.html (hunk 11 above)
Comment (johangiraud) [reply cafebabel/templates/_social-networks-macro.html]: hum ok I get your point but it doesn't work your way. If I add text content, I can't have them hidden so far. Any ideas?

Comment (davidbgk) [reply cafebabel/templates/_social-networks-macro.html]: At least use a `title` attribute, see https://twitter.com/stevefaulkner/status/978310635023949830

File: cafebabel/static/js/script.js [hunk 12]
```diff
@@ -26,6 +26,36 @@ Array.from(document.querySelectorAll('a')).forEach(a => {
...
+    elem[style] = from + step * (to - from)
+    if (step === 1) clearInterval(timer)
+  })
+  elem[style] = from
+}
+const subscribeLink = document.querySelector(
+  '#social-networks .icon-newsletter'
+)
+subscribeLink.addEventListener('click', () => {
+  const target = document.querySelector('#static-pages-links')
```
Comment (davidbgk) [inline cafebabel/static/js/script.js]: It should be retrieved from the `event` like `const target = document.querySelector(event.target.getAttribute('href'))`

Comment (vinyll) [reply cafebabel/static/js/script.js]: This smooth scroll doesn't work for me.
It just goes to the anchor as normal.
I tried prepending with `event.preventDefault()` to cut off the default behavior (that I suppose we need to turn off) and it wouldn't move at all anymore.

Comment (johangiraud) [reply cafebabel/static/js/script.js]: ah oui, on firefox, it doesn't work... desperate..

File: cafebabel/templates/_social-networks-macro.html (hunk 11 above)
Comment (johangiraud) [reply cafebabel/templates/_social-networks-macro.html]: I did that cause I couldn't have it done fast enough with `text-indent: 9999px blabla` 😞

File: cafebabel/templates/_social-networks-macro.html [hunk 13]
```diff
@@ -1,11 +1,12 @@
...
-    <li><a href={{ social_network_url_for('instagram') }} class=icon-instagram></a></li>
-    <li><a href={{ social_network_url_for('youtube') }} class=icon-youtube></a></li>
+    <li><a href={{ social_network_url_for('facebook') }} class=icon-facebook title="Cafébabel on Facebook"></a></li>
+    <li><a href={{ social_network_url_for('twitter') }} class=icon-twitter title="Cafébabel on Twitter"></a></li>
+    <li><a href={{ social_network_url_for('instagram') }} class=icon-instagram title="Cafébabel on Instagram"></a></li>
+    <li><a href=#static-pages-links class=icon-newsletter title="Cafébabel on Facebook"></a></li>
+    <li><a href={{ social_network_url_for('youtube') }} class=icon-youtube title="Cafébabel on Facebook"></a></li>
     {% if with_linkedin %}
-        <li><a href={{ social_network_url_for('linkedin') }} class=icon-linkedin></a></li>
+        <li><a href={{ social_network_url_for('linkedin') }} class=icon-linkedin title="Cafébabel on Facebook"></a></li>
```
Comment (davidbgk) [inline cafebabel/templates/_social-networks-macro.html]: Wrong `title`

File: cafebabel/templates/_social-networks-macro.html (hunk 11 above)
Comment (vinyll) [reply cafebabel/templates/_social-networks-macro.html]: I guess the title fallback could be a minimum with minimal time consumption in that case.
This applies to all links that are empty.

File: cafebabel/templates/_social-networks-macro.html (hunk 13 above)
Comment (johangiraud) [reply cafebabel/templates/_social-networks-macro.html]: 😮 https://github.com/cafebabel/cafebabel.com/pull/386/commits/410f43a4f781a722c51055df4c2557b055c1c01b

File: cafebabel/static/js/script.js (hunk 12 above)
Comment (johangiraud) [reply cafebabel/static/js/script.js]: Fixed here @vinyll ebc33fc 😅

File: cafebabel/templates/_social-networks-macro.html (hunk 11 above)
Comment (johangiraud) [reply cafebabel/templates/_social-networks-macro.html]: ok !!

**Comment by johangiraud** [comment]
To me it's all good.

File: cafebabel/templates/_social-networks-macro.html [hunk 14]
```diff
@@ -1,11 +1,12 @@
 {% macro social_networks(with_linkedin=False) %}
   <ul>
-    <li><a href={{ social_network_url_for('facebook') }} class=icon-facebook></a></li>
-    <li><a href={{ social_network_url_for('twitter') }} class=icon-twitter></a></li>
-    <li><a href={{ social_network_url_for('instagram') }} class=icon-instagram></a></li>
-    <li><a href={{ social_network_url_for('youtube') }} class=icon-youtube></a></li>
+    <li><a href={{ social_network_url_for('facebook') }} class=icon-facebook title="Cafébabel on Facebook"></a></li>
+    <li><a href={{ social_network_url_for('twitter') }} class=icon-twitter title="Cafébabel on Twitter"></a></li>
+    <li><a href={{ social_network_url_for('instagram') }} class=icon-instagram title="Cafébabel on Instagram"></a></li>
+    <li><a href=#static-pages-links class=icon-newsletter title="Cafébabel on Facebook"></a></li>
```
Comment (davidbgk) [inline cafebabel/templates/_social-networks-macro.html]: I think that title is wrong 😉

**Review by davidbgk — APPROVED** [review]
With comment.

Note that it's not smooth at all with Firefox for me but it doesn't alter the feature. I already had display issues with that big footer though… like when you scroll quickly or using keyboard, sometimes I'm afraid it is altering the user experience while only adding some fanciness but I'm bad at judging (in that domain) 😄

File: cafebabel/templates/_social-networks-macro.html (hunk 14 above)
Comment (johangiraud) [reply cafebabel/templates/_social-networks-macro.html]: rooo https://github.com/cafebabel/cafebabel.com/pull/386/commits/a26d84cb31680b7f0dfb19db76d0580a82af6628

**Comment by johangiraud** [comment]
I know. I can't have it work as I would like. I set a tiny duration cause there is a short jump on click. With longer duration, you actually notice this "jumping effect" a lot more even though the scroll is way smoother. That's the first time I do it with pure js and I don't know how to improve that. Suggestions are welcome :-) @davidbgk

**Comment by davidbgk** [comment]
@cafebabel-webmaster stop messing with browsers' native scrolls and anchors 😉

**Pull request merged**